README.md
¡ Os presento mi proyecto personal ! Se trata de un editor de código online colaborativo :goodjob: 

Esta en fase _Alpha_ :PatrikBaby:  así que la parte visual que tiene es solo algo temporal

*¿ En qué consiste ?*
Es una web para prototipar de forma rápida webs (frontend de momento sólo) con amigos, tan fácil como entrar y darle a crear sala, ¡ sin necesidad de crear usuarios ! :feelsgoodman: 
Para entrar en la sala de tu amigo sólo hay que darle a join y introducir el mismo nombre de la sala :WhatIstThat: 

* Futuras mejoras *
- Más lenguajes: Compilador de C, C++, Processing ( ya hecho, falta merge ), NodeJS, PHP. (Y alguno más que me sugiráis ^.^)
- Implementación de repositorios: Integrar de forma interna una forma de utilizar repositorios de github y realizar commits desde dentro de la web
- Sistema de capetas / proyecto: De momento solo soporta trabajar con 1 archivo a la vez
- Algún tipo de persistencia de sala / código:  El servidor detecta cuando usa sala ha quedado vacía y la borra para no malgastar memoria, posible implementación de usuario para tener un dashboard o algo parecido no se que hacer de momento :rofl: 

Agradezco cualquier feedback de como mejorar la web y qué características serian interesantes :heart_eyes: .

La web se puede usar sin problemas y esta abierta a quien la necesite en su versión alpha :Messirve: .

Processing/changes.md
# Processing 4.0 beta 7

*Revision 1282 – 3 March 2022*

The main thing for this release is to prevent an issue where code might be deleted (!) due to the change in beta 6.

…it also rolls back a change that broke Python Mode. Sorry, snake people!

…but also: a new feature! A very cool new feature! See below.


## Regression Repair

* Bring back `getMainProgram()` for Python Mode. [#409](https://github.com/processing/processing4/issues/409)

* Fix an issue where opening a single .pde file in an unrelated folder (i.e. Downloads) would delete the entire folder when deleting the "sketch" without being 100% obvious what it was doing. [#424](https://github.com/processing/processing4/issues/424)

* Andrés tracked down the problem with `POINTS` and `PShape` that was introduced in 4.0 beta 3. [#345](https://github.com/processing/processing4/issues/345), [#432](https://github.com/processing/processing4/pull/432)


## I'm just here for the ratio

Added a new `windowRatio(width, height)` function that remaps screen and mouse coordinates to the specified aspect ratio. 

The general idea is that you set a ratio, and whether you're using `fullScreen()` or `setResizable(true)`, the sketch will scale its coordinates (and the coordinates of the mouse to fit that same ratio. 
  
For instance, use `windowRatio(1280, 720)` in your code, and then all your coordinates will be re-mapped to that range—but always keeping the aspect ratio. Probably easiest to see by running this sketch:

```processing
void setup() {
  windowResizable(true);
  windowRatio(1280, 720);
  
  cursor(CROSS);
  strokeWeight(10);
}

void draw() {
  background(240);
  fill(255);
  rect(0, 0, rwidth, rheight);
  
  fill(0);
  textAlign(CENTER, CENTER);
  textSize(200);
  text(rmouseX + ", " + rmouseY, rwidth/2, rheight/2);
}
```

When using `windowRatio()`, the new `rwidth` and `rheight` variables contain the width and height that were passed to `windowRatio()` and can be used in place of `width` and `height` elsewhere in your code. 

Similarly, `rmouseX` and `rmouseY` contain the mouse position, scaled by the current ratio.

If your window is taller (or wider) than the specified ratio, your sketch will be moved (using `translate()` to fit the space, and `rmouseX` and `rmouseY` might even have negative values if they're outside the box. (Again, try running the code above.) 

The exact amount of space that's extra can be found in the `ratioLeft` and `ratioTop` variables. And the current `scale()` being used for the ratio can be found in `ratioScale`. 

Mind you, it might be tempting to use windowRatio(16, 9), but keep in mind that means your horizontal coordinates will be numbers between 0 and (almost) 16, and vertical will be between 0 and 9. Not a great way to work! (And at some point, likely to introduce quirks as you learn about the inaccuracies of floating point number accuracy.)


## Minor Mites

* Half a dozen other fixes/changes for opening sketches and how it identifies the Mode and the rest. If you notice anything weird, let us know asap.

* Rewrote our native code for going full screen on macOS, retiring jAppleMenuBar after a decade of service. Also gets things working on Apple Silicon.

* Prevent “You must first install tweak Mode to use this sketch” when using sketches that date back to 2.0. [#415](https://github.com/processing/processing4/issues/415)

* Swap more uses of `Util.deleteFile()` and move files to the Trash (or Recycle Bin) instead. This can make more mess, but it reduces the chances of something bad happening when Processing deletes a file.

* Change straight quotes to smart quotes in the `PDE.properties` file. This just means that more UI and error messages will have nicer quotes.


# Processign 4.0 beta 6

*Revision 1281 – 13 February 2022*

Continuing the unintentional tradition of Processing releases on [Super Bowl Sundays](https://en.wikipedia.org/wiki/Super_Bowl_Sunday), here's beta 6.


## Scaling and stretching on Windows

* Windows scaling at 125% was stretching sketches to make them look crusty. [#378](https://github.com/processing/processing4/issues/378)

* `pixelDensity()` not working in exported Windows applications. [#5414](https://github.com/processing/processing/issues/5414#issuecomment-841088518)

* Implement sketch scaling in `PApplet` so that they take UI sizing into account. [#4897](https://github.com/processing/processing/issues/4897), [#4894](https://github.com/processing/processing/issues/4894)

These changes also include some extra native code on Windows, which is described in detail [here](https://github.com/processing/processing4/tree/master/build/windows/fenster).


## Hell freezes over

For those of you that use Git (or other version control) this release includes an experimental(!) feature that lets you separate the name of the sketch from the name of the main tab. This means that you can use "Save As…" and it won't rename the file for the main tab. *Or* you can rename a sketch folder and the sketch inside will still work properly.

Uncheck the “Keep sketch name and main tab in sync” option in Preferences, and you're on your way. This is a *major* change to some of the oldest code in the PDE. If it has bugs, it may eat code. **Use with caution and report any issues you see immediately!**

Again, please help test this feature and [file issues](https://github.com/processing/processing4/issues/new) if you run into trouble.


## Bug fixes

* Actually set the correct names for the `windowMove()` and `windowResize()` functions that were announced in beta 5. [#53](https://github.com/processing/processing4/issues/53)

* PDF export not working with Processing 4.0 beta 5. [#395](https://github.com/processing/processing4/issues/395)

* Lots of issues discovered when making the sketch naming change. For instance, it looks like it was sometimes necessary to twice select the Mode for Android sketches. Also sorts out issues with when `sketch.properties` was or was not written.


index.js
const path = require('path');
const express = require('express');
const http = require('http');
const socketio = require('socket.io');
// const { randomInt } = require('crypto');
const child_process = require('child_process');
const fs = require('fs');
const fs_Extra = require('fs-extra');
const kill = require('tree-kill');

const app = express();
const server = http.createServer(app);
const PORT = process.env.PORT || 5000;
const io = socketio(server);

const frame_rate = 144;
const premium_frame_rate = 60;

const init_iframe_code = {
    html: ["<h1>", '\tHola Jaime', "</h1>"].join("\n"),
    css: ["h1 {", '\tcolor: red;', "}"].join("\n"),
    js:  ["function x() {", '\tconsole.log("Hola Jaime!");', "}", "x();"].join("\n")
};
const init_processing_code = {
    processing: '',
}

// heroku logs: heroku logs -t --app code-jaime-online
// heroku log error: heroku logs --tail 500 --app code-jaime-online o heroku logs -n 500 --app code-jaime-online

// AUX
let debounceTimer;
function debounce (callback, time) {
  clearTimeout(debounceTimer);
  debounceTimer = setTimeout(callback, time);
};

// defining processing init_processing_code getting from template.pde
function get_init_processing() {
    // read temp/template/template.pde code and change processing code
    fs.readFile(path.join(__dirname, 'temp', 'template', 'template.pde'), function(err, data) {
        if (err) throw err;
        init_processing_code.processing = data.toString();
    });
}

get_init_processing();

// Seteando carpeta estatica, carpeta donde contiene todos los datos que requiere el usuario cuando hace la peticion
// a la web buscando recursos.
app.use(express.static(path.join(__dirname, 'dist')))

console.log("Empezando!!");

const iframe_code_rooms = {};
const socket_rooms = {};
const socket_users = {};

// Funcion que se ejecuta cuando un usuario se conecta al websocket
io.on('connection', (socket) => {
    console.log("Nueva conexion!!");

    socket.on("disconnecting", () => {
        // Deleting users
        if (socket_users[socket.id] == null) return; // check if user exists
        console.log('user exists');
        var room_id = socket_users[socket.id].room;
        if (socket_rooms[room_id] == null) return; // check if room exists
        console.log('room exists');
        if (!socket_rooms[room_id].includes(socket.id)) return;  // check if user is in room
        console.log('user inside room');
        
        // Deleting user form rooms if was in one
        socket_rooms[room_id].splice(socket_rooms[room_id].indexOf(socket.id), 1);
        // remove user from users
        delete socket_users[socket.id];
        
        // if room is empty, delete room
        if (socket_rooms[room_id].length == 0) {
            
            if (socket_rooms[room_id].program == 'processing') {
                // Delete folder of the room
                fs.rmSync(socket_rooms[room_id].dir, { recursive: true, force: true });
            }

            
            delete socket_rooms[room_id];
        }

        // DEBUG
        // console.log("**** DISCONNECTING ****");
        // console.log(socket.id);
        // console.log(socket_users);
        // console.log(socket_rooms);
        // console.log("**********************");
    });   

    // msg to change code on server and other clients in the same room
    socket.on('client_code', (data) =>{
        if (socket_users[socket.id] == null){
            socket.emit('error', {
                code: 408,
                info: 'User not found'
            });
            return; // check if user exists
        }
        socket.broadcast.to(socket_users[socket.id].room).emit('server_code', data);
        
        if (data.id !== 'html' && data.id !== 'css' && data.id !== 'js' && data.id !== 'processing') return; // check if id is valid
        
        iframe_code_rooms[socket_users[socket.id].room][data.id] = data.code; // update iframe code of the room
        
    });
    

    // msg to set up client code at the beginning
    socket.on('get_code_client', (data) => {
        if (socket_users[socket.id] == null) { // check if user exists
            socket.emit('error', {
                code: 408,
                info: 'User not found'
            });
            return; // check if user exists
        }

        socket.emit('get_code_server', iframe_code_rooms[socket_users[socket.id].room]);
    });

    socket.on('create_room', data => {
        console.log(data);
        // check if data room is valid
        if (data.room == null || data.room == "") {
            socket.emit('error', {
                code: 400,
                info: 'Room name is invalid'
            });
            return;
        }

        // check if room exists
        if (socket_rooms[data.room] != null) {
            socket.emit('error', {
                code: 402,
                info: 'Room already exists'
            });
            return;
        }

        // check if program setted
        if (data.program !== 'web' && data.program !== 'processing') {
            socket.emit('error', {
                code: 405,
                info: 'Program is invalid ( Web | Processing )',
                message: data
            });
            return;
        }

        // check if valid name for processing
        if (data.program !== 'processing' && data.room == 'template') { // TODO: should migrate template folder to another name
            socket.emit('error', {
                code: 406,
                info: 'Room already exists',
                message: data
            });
            return;
        }

        //check if valid user type
        if (data.user_type !== 'user' && data.user_type !== 'img_user') {
            socket.emit('error', {
                code: 407,
                info: 'User type is invalid, refresh the page and try again',
                message: data
            });
            return;
        }

        // create room and join the user
        socket_rooms[data.room] = [];
        socket_rooms[data.room].push(socket.id);
        socket_rooms[data.room].program = data.program;
        
        socket_users[socket.id] = {
            room: data.room,
            user_type: data.user_type
        };
        
        // *** FOR WEB ***
        console.log('Setting up iframe code');
        iframe_code_rooms[data.room] = {};
        if (data.program == 'web') {
            // create initial code
            Object.assign(iframe_code_rooms[data.room], init_iframe_code);
            console.log('Setting up web code');
            console.log(init_iframe_code);
        }
        // *** FOR PROCESSING ***
        else if (data.program == 'processing') {
            // create initial code
            // create a copy of init_processing_code
            Object.assign(iframe_code_rooms[data.room], init_processing_code);
            
            // *** this is moved to client side ***
            // console.log('Setting up processing code');
            // console.log(init_processing_code);
            
            // console.log('Creating folder for the room');
            // var dir_template = path.join(__dirname, 'temp', 'template');
            // var dir_room = path.join(__dirname, 'temp', data.room);
            // cloneFolder(dir_template, dir_room, data.room);
            
            // socket_rooms[data.room].dir = dir_room;
            // console.log('Room folder created');
        }
        
        // join the room
        socket.join(data.room);

        // send response to client
        socket.emit('enter_room', {
            room: data.room,
            program: data.program
        });
    });

    socket.on('join_room', data => {
        // check if data room is valid
        if (data.room == null || data.room == "") {
            socket.emit('error', {
                code: 400,
                info: 'Room name is invalid'
            });
            return;
        }

        // check if room exists
        if (socket_rooms[data.room] == null) {
            socket.emit('error', {
                code: 401,
                info: 'Room does not exists'
            });
            return;
        }

        // join the user
        socket_rooms[data.room].push(socket.id);
        socket_users[socket.id] = {
            room: data.room,
            user_type: data.user_type
        };

        console.log('New user!');
        console.log(data);

        socket.join(data.room);
          // send response to client
          socket.emit('enter_room', {
            code: 200,
            room: data.room,
            program: socket_rooms[data.room].program
        });
    });
    
    socket.on('run_code_client', data => {
        if (socket_users[socket.id] == null) return;
        
        var room_name = socket_users[socket.id].room;

        if (socket_rooms[room_name].program != 'processing') {
            socket.emit('error', {
                code: 403,
                info: 'Program is not processing'
            });
            return;
        };
        
        // check if img_user is in the room
        var img_user_room_id = null;
        for (var i = 0; i < socket_rooms[room_name].length; i++) {
            if (socket_users[socket_rooms[room_name][i]].user_type == 'img_user') {
                img_user_room_id = socket_rooms[room_name][i];
                console.log('break');
                break;
            }
        }
        console.log('id img_user_room: ' + img_user_room_id);

        if (img_user_room_id == null) {
            socket.emit('error', {
                code: 409,
                info: 'There is no client running the code in the room'
            });
            console.log(socket_rooms);
            console.log(socket_users);
            return;
        }

        // create a random string
        console.log('running code');

        // send code to img_user only
        io.to(img_user_room_id).emit('run_code_server', {
            code: data.code,
        });

        // var room_name = socket_users[socket.id].room;
        // run_code(socket, room_name);
    });

    socket.on('stop_code_client', () => {
        if (socket_users[socket.id] == null) return;
        socket.broadcast.to(socket_users[socket.id].room).emit('stop_code_server');
    });
    
    // redirecting image from img_user to clients connected to the same room
    socket.on('image_client', img => {
        if (socket_users[socket.id] == null) return;
        socket.broadcast.to(socket_users[socket.id].room).emit('image_server', img);
    });
    
    // redirecting output from img_user to clients connected to the same room
    socket.on('processing_output_client', output => {
        if (socket_users[socket.id] == null) return;
        socket.broadcast.to(socket_users[socket.id].room).emit('processing_output_server', output);
    });
});

const interval_time = 1000/frame_rate;

function cloneFolder(source, target, room_name) {

    // check if folder exists and create it if not
    if (!fs.existsSync(target)){
        fs.mkdirSync(target, { recursive: true });
    }

    // copy files from source to target
    fs_Extra.copy(source, target, function(error) {
        if (error) {
            throw error;
        } else {
          console.log("success!");
          // rename template.pde to (room name).pde
            fs.renameSync(path.join(target, 'template.pde'), path.join(target, room_name + '.pde'));
          console.log('room name: ' + room_name);
        }
    });
}

function run_script(command, args, cwd, time_out=null, callback) {
    console.log("Starting Process.");
    console.log("Command: " + command);
    console.log("Args: " + args);
    console.log("CWD: " + cwd);
    console.log("Timeout: " + time_out);

    if (time_out == null) {
        console.log("Error no time setted of process child.");
        return;
      }
    
    var child = child_process.spawn(command, args, cwd);
    console.log("Child created");
    
    child.stdout.setEncoding('ascii');
    child.stdout.on('data', function(data) {
        callback(data);
    });

    child.stderr.setEncoding('ascii');
    child.stderr.on('data', function(data) {
        console.log("Error: " + data);
    });

    child.on('close', function() {
        console.log('finish code: ' + child.exitCode);
    });

    // setTimeout( () => {
    //     kill(child.pid);
    //     console.log("Process killed.");
    // }, time_out);

    return child;
}

server.listen(PORT, () => {console.log(`runing on port ${PORT}`);});

function run_code(socket, room_name){
    var time_out = 10000; // 10 seconds
    console.log('running on: ', process.platform);

    if (process.platform === 'win32') {
        console.log('running on windows');
        var child = run_script(path.join(__dirname, 'Processing', 'win32', 'processing-java.exe'), ["--force", `--sketch=${path.join(__dirname, 'temp', room_name)}`, `--output=${path.join(__dirname, 'temp', room_name,'out')}`, "--run"], {cwd:`${path.join(__dirname, 'temp', room_name)}`}, time_out, function(buf) {
            socket.emit('processing_output_server', buf);
        });
    }
    if (process.platform === 'linux') {
        console.log('running on linux');
        var child = run_script(path.join(__dirname, 'Processing', 'linux', 'processing-java'), ["--force", `--sketch=${path.join(__dirname, 'temp', room_name)}`, `--output=${path.join(__dirname, 'temp', room_name,'out')}`, "--run"], {cwd:`${path.join(__dirname, 'temp', room_name)}`}, time_out, function(buf) {
            socket.emit('processing_output_server', buf);
        });
    }
        
    if (child == null)  return;

    // TODO: DISPLAY IMAGE, set it to room attribute so it can stopped later on
    var int_display = setInterval(() => {
        // read frame.svg from template/img_output
        fs.readFile(path.join(__dirname, 'temp', room_name, 'img_output', 'frame.svg'), function(err, data) {
            if (err) {
                console.log('Error reading file: ' + err);
                // if still not sending image reset all finish timers, so it's time_out when start sending image to the client
                debounce(() => {
                    // kill process
                    kill(child.pid);
                    console.log("Process killed.");

                    // stop sending frames to client
                    clearInterval(int_display);
                    console.log('display cleared');
                    socket.emit('processing_info_server', 'Processing finished');
                    
                }, time_out);
                
                // clearInterval(int_display);
                // console.log('display cleared becouse error occured');
                // stop interval
                return;
            }
            // covert to base64
            var base64data = new Buffer(data).toString('base64');
            // console.log('sent');
            socket.emit('image_server', base64data);
        });
        // emit image to socket
    }, interval_time);

    
}
